source/opt/value_number_table.cpp

```diff
@@ -48,24 +58,115 @@ uint32_t ValueNumberTable::GetValueNumber(spvtools::ir::Instruction* inst) {
   // read only.  However, if this is ever relaxed because we analyze stores, we
   // will have to add a new case for volatile loads.
   if (inst->IsLoad() && !inst->IsReadOnlyLoad()) {
-    uint32_t value_number = TakeNextValueNumber();
-    id_to_value_[inst->result_id()] = value_number;
-    return value_number;
+    value = TakeNextValueNumber();
+    id_to_value_[inst->result_id()] = value;
+    return value;
+  }
+
+  // When we copy an object, the value numbers should be the same.
+  if (inst->opcode() == SpvOpCopyObject) {
+    value = GetValueNumber(inst->GetSingleWordInOperand(0));
+    if (value != 0) {
+      id_to_value_[inst->result_id()] = value;
+      return value;
+    }
   }
 
+  // Phi nodes are a type of copy.  If all of the inputs have the same value
+  // number, then we can assign the result of the phi the same value number.
+  if (inst->opcode() == SpvOpPhi) {
+    value = GetValueNumber(inst->GetSingleWordInOperand(0));
+    if (value != 0) {
+      for (uint32_t op = 2; op < inst->NumInOperands(); op += 2) {
+        if (value != GetValueNumber(inst->GetSingleWordInOperand(op))) {
+          value = 0;
+          break;
+        }
+      }
+      if (value != 0) {
+        id_to_value_[inst->result_id()] = value;
+        return value;
+      }
+    }
+  }
+
+  // Replace all of the operands by their value number.  The sign bit will be
+  // set to distinguish between an id and a value number.
+  ir::Instruction value_ins(context(), inst->opcode(), inst->type_id(),
+                            inst->result_id(), {});
+  for (uint32_t o = 0; o < inst->NumInOperands(); ++o) {
+    const ir::Operand& op = inst->GetInOperand(o);
+    switch (op.type) {
+      case SPV_OPERAND_TYPE_ID: {
```

Comment:
This misses scope ID and memory semantics ID.  Prefer  if(spvOpIsIdTytpe(op.type)), which works here because you're already skipping past result-id and type-id (which is SPV_OPERAND_TYPE_TYPE_ID)

Reply:
Done.